### DISCUSSION TITLE
how to improve performance?

### DISCUSSION THREAD

**Discussion by TITC** [lukas-blecher/LaTeX-OCR#146] [General]
This discussion is more close to an endless answer's question.

## Recap

Here is some related information I have found, contains two issues https://github.com/lukas-blecher/LaTeX-OCR/issues/27#issue-948197787, https://github.com/lukas-blecher/LaTeX-OCR/issues/44#issue-1033177422 about BLEU and one [discussion](https://github.com/lukas-blecher/LaTeX-OCR/discussions/11) about LR schedule and optimizer.

## Experiment
I have added 0.35 million Wikipedia data to the training set, so there are 0.5 million in total. After that, I create a wandb sweep.

### panel1  [date&bleu](https://wandb.ai/yht4work/LaTeX-OCR-pix2tex/reports/Shared-panel-22-05-05-11-05-64--VmlldzoxOTQ5NDg1?accessToken=w50bbctb4b4h2b5jw7nsd1jvaf9vb6un411nkwcgoel4uaqw80ngw61vlasy6bfa)
![image](https://user-images.githubusercontent.com/35098797/166858235-eaeedac2-1cf8-440e-9fc1-3cf468ca194b.png)

### panel2 [hyperparamer influence](https://wandb.ai/yht4work/LaTeX-OCR-pix2tex/reports/Shared-panel-22-05-05-11-05-27--VmlldzoxOTQ5NTMw?accessToken=u3q53jmfsv9qzlg5pgio2vy7kb1o5se10d01lnk1b1xflob3l2f47ef6566qd9g2)
![image](https://user-images.githubusercontent.com/35098797/166858368-819a1911-f147-4692-bc99-cbcae73c8080.png)

### panel3 [fully connected layer style](https://wandb.ai/yht4work/LaTeX-OCR-pix2tex/reports/Shared-panel-22-05-05-11-05-48--VmlldzoxOTQ5NTY4?accessToken=rdfqemmescoqj6qtekg70v5gb4sxc6spoz8kkgi2t5iw77qlr081389yxsf4gy1i)

![image](https://user-images.githubusercontent.com/35098797/166859576-599150c9-77c6-4ca9-a823-62c3d7fb2014.png)

### panel4 [token acc](https://wandb.ai/yht4work/LaTeX-OCR-pix2tex/reports/Shared-panel-22-05-05-11-05-53--VmlldzoxOTQ5NTgx?accessToken=79pnx3e3o2epunjo15rnq29akd0v0iogtukkgs95e95ph0jaopy36oa287t2w636)

![image](https://user-images.githubusercontent.com/35098797/166859947-7ba17e16-766f-4628-90bf-48df8b82c37c.png)

### panel5 [val_bleu](https://wandb.ai/yht4work/LaTeX-OCR-pix2tex/reports/Shared-panel-22-05-05-11-05-30--VmlldzoxOTQ5NTg1?accessToken=97k8p6d335yoyjk4b6bjz73fu9nl6oveh6m01v6mma5ew3t5a5cprmxgledcdub7)

![image](https://user-images.githubusercontent.com/35098797/166860111-cfe12258-202f-4a52-aa59-4bf4f27173a3.png)

---
I have added more data and tried different hyperparameters which include lr schedule and optimizer, even the number of encoder layers and decoder layers, see panel3 [fully connected layer style](https://wandb.ai/yht4work/LaTeX-OCR-pix2tex/reports/Shared-panel-22-05-05-11-05-48--VmlldzoxOTQ5NTY4?accessToken=rdfqemmescoqj6qtekg70v5gb4sxc6spoz8kkgi2t5iw77qlr081389yxsf4gy1i) for more detail. But nothing gets better. What can I do to achieve higher performance? 

 5 days before, I reckon more data and hyper-parameter searching will work, but now the result is frustrating.

**Comment by lukas-blecher** [comment]
One thing I noticed that I don't know what it means is the `batchsize` parameter in the `fully` connected diagram. The values reported are not realistic. What is the meaning of that? Also, the batch size shouldn't have such an impact on the performance of the model.

I don't believe that the optimizer should be one of the first things to swap out. I could be wrong but the sentiment in ML seems to be that Adam or AdamW are the way to go. Also `min/max_height/width` shouldn't be parameters to change in the sweep because that means every model trains on a different dataset. Similarly, the `temperature` should be equal for all runs of the sweep so that the evaluation metrics can be compared with each other.

Instead you could play around with the `backbone_layers` in number of layers and depth.

I also don't understand the range of `gamma`. I'm using it for the StepLR scheduler and there it is used as a multiplicative for the lr (not sure what happens in your code for other schedulers), hence the value should be <1 if you want the LR to decrease. 

The metrics have a high variance, what value for `valbatches` did you use? Could also be connected to the `temperature` field as well.

Lastly, you can try to train for more iterations for the best hyper parameters. I've trained one model for 300k steps and it still improved somewhat in the last half.

**Reply by TITC** [reply]
Wow, a lot of problems of mine exposed to sunshine.

---
1. batch size 
```yaml
  batchsize:
    distribution: q_log_uniform
    max: 50
    min: 5
    q: 8
```
about [q_log_uniform](https://docs.wandb.ai/guides/sweeps/configuration), the official doc described as below
> Quantized log uniform. Returns round(X / q) * q where X is log_uniform. q defaults to 1.

2. sole variable
The reason I'm haven't control variates each time is that I think wandb `bayes` search strategy will config this automatically. The sweep code has nothing different with [train.py](https://github.com/lukas-blecher/LaTeX-OCR/blob/main/pix2tex/train.py) except removing argparse.
```python
    parser = argparse.ArgumentParser(description='Train model')
    parser.add_argument('--config', default=None,
                        help='path to yaml config file', type=str)
    parser.add_argument('--no_cuda', action='store_true', help='Use CPU')
    parser.add_argument('--debug', action='store_true', help='DEBUG')
    parser.add_argument(
        '--resume', help='path to checkpoint folder', action='store_true')
    parsed_args = parser.parse_args()
    if parsed_args.config is None:
        with in_model_path():
            parsed_args.config = os.path.realpath('settings/debug.yaml')
    with open(parsed_args.config, 'r') as f:
        params = yaml.load(f, Loader=yaml.FullLoader)
    args = parse_args(Munch(params), **vars(parsed_args))
    logging.getLogger().setLevel(logging.DEBUG if parsed_args.debug else logging.WARNING)
    seed_everything(args.seed)
    if args.wandb:
        if not parsed_args.resume:
            args.id = wandb.util.generate_id()
        wandb.init(config=dict(args), resume='allow',
                   name=args.name, id=args.id)
    train(args)
```
after
```python
    kwargs = {}
    kwargs["device"] = "cuda"
    kwargs["debug"] = False
    kwargs["resume"] = False
    kwargs["no_cuda"] = False
    kwargs["config"] = "model/settings/config.yaml"
    with open(kwargs["config"], 'r') as f:
        params = yaml.load(f, Loader=yaml.FullLoader)
    args = parse_args(Munch(params), **kwargs)
    wandb.init(config=dict(args), name=args.name)
    train(args)
```


3. backbone_layers

I will have a try and could you give some suggestions about a reasonable variable range?

4. gamma
I just create the sweep followed by [this tutorial](https://github.com/wandb/examples/tree/master/examples/pytorch/pytorch-cnn-fashion). There a lot of variables I do not actually know what's meaning, just create it and run. After your careful explanation, I get more clear about that part. Thanks for you explanation.

5. valbatches is the default 100 same with [config.yaml](https://github.com/lukas-blecher/LaTeX-OCR/blob/main/pix2tex/model/settings/config.yaml#L51)
6. sweep.config
```yaml
method: bayes
metric:
  goal: maximize
  name: val/bleu
parameters:
  batchsize:
    distribution: q_log_uniform
    max: 50
    min: 5
    q: 8
  dim:
    distribution: int_uniform
    max: 512
    min: 128
  encoder_depth:
    distribution: int_uniform
    max: 16
    min: 2
  gamma:
    distribution: uniform
    max: 1.999
    min: 0.49975
  heads:
    distribution: int_uniform
    max: 16
    min: 4
  lr:
    distribution: uniform
    max: 0.002
    min: 0.0005
  lr_step:
    distribution: int_uniform
    max: 60
    min: 15
  max_height:
    distribution: int_uniform
    max: 384
    min: 96
  max_seq_len:
    distribution: int_uniform
    max: 1024
    min: 256
  max_width:
    distribution: int_uniform
    max: 1344
    min: 336
  min_height:
    distribution: int_uniform
    max: 64
    min: 16
  min_width:
    distribution: int_uniform
    max: 64
    min: 16
  num_layers:
    distribution: int_uniform
    max: 16
    min: 2
  optimizer:
    distribution: categorical
    values:
      - Adam
      - SGD
      - Adadelta
      - RMSprop
      - Adamax
      - AdamW
      - Adagrad
      - SparseAdam
      - ASGD
      - LBFGS
      - NAdam
      - RAdam
      - Rprop
  patch_size:
    distribution: int_uniform
    max: 32
    min: 8
  scheduler:
    distribution: categorical
    values:
      - StepLR
      - LambdaLR
      - MultiplicativeLR
      - MultiStepLR
      - LinearLR
      - OneCycleLR
      - ConstantLR
      - ExponentialLR
      - CosineAnnealingLR
      - ChainedScheduler
      - SequentialLR
      - ReduceLROnPlateau
      - CyclicLR
      - CosineAnnealingWarmRestarts
  temperature:
    distribution: uniform
    max: 0.9
    min: 0.1
program: sweep.py
```
Is there any strange setting in this config? I will remove `gamma` and `optimizer`, `batchsize` and train more iterations.

**Reply by TITC** [reply]
awkward, those parameters never work because I forgot to cover `args` parameter

```python
args = Munch(wandb.config)
```

so those figures all run at the same hyperparameter ಥ_ಥ

**Reply by TITC** [reply]
> Instead you could play around with the `backbone_layers` in number of layers and depth.

what's the depth refer to? the  `encoder_depth` or `backbone_layers` depth? I noticed add the number of  `backbone_layers`, it will also need to increase patch size. For example, [2, 3, 7] to [17, 4, 16, 16] need to multiply 32to patch size.
```shell
AssertionError: patch_size needs to be multiple of 32 with current backbone configuration
```

**Reply by lukas-blecher** [reply]
Ok, so then you will also need to be smarter about the LR scheduler args, because they are named and specific to each scheduler. That's kinda hard to do automatically. Maybe do independent sweeps for a couple of scheduler?

> Is there any strange setting in this config? I will remove gamma and optimizer, batchsize and train more iterations
Also don't forget to fix the `temperature`.





> awkward, those parameters never work because I forgot to cover `args` parameter
> 
> ```python
> args = Munch(wandb.config)
> ```
> 
> so those figures all run at the same hyperparameter ಥ_ಥ

ok that sucks ^^
But also kinda weird that there was such a difference in performance. Can't really explain that. Especially because the seed was the same.
I thought this covers everything: https://github.com/lukas-blecher/LaTeX-OCR/blob/084fb035c421cb928e8a580901d4369a28306b13/pix2tex/utils/utils.py#L37-L49


> what's the depth refer to? the encoder_depth or backbone_layers depth?

The depth is for the backbone layers. You can customize the number of layers by adding/removing list items of `backbone_layers` as you've done.
The depth is used here:
https://github.com/rwightman/pytorch-image-models/blob/75144395739162a153dae5628320b85b5895634b/timm/models/resnetv2.py#L280

The error should have told you that for 4 layers the ViT patches are made out of at least 32x32 pixels but also accepts all higher powers of 2: 64, 128, ...
I do notice, that that's not the same as what the error says, but I think I wrote the wrong message in the code.

**Reply by TITC** [reply]
> But also kinda weird that there was such a difference in performance. Can't really explain that. Especially because the seed was the same.

I think the reason is I removed that line.
```python
    seed_everything(args.seed)
```
---

Hope I will have some valuable news to share with you over the next 42 days.

**Reply by TITC** [reply]
Here is an [excel ](https://1drv.ms/x/s!AlNM4pqY6ecSlssTeYs0Vmgy2I2Rgg?e=uEZdbb)file I am going to record training data in it. Leave a message to share information.

**Reply by lukas-blecher** [reply]
That already looks promising! 
Quick question: What does "e&s" mean (e10s980+e3s6217+e1s6999)?
Also: I like to look at the loss with a logarithmic scale and some smoothing. I think that way it's more obvious if something is happening.

**Reply by TITC** [reply]
for the append epoch, I choose the best checkpoints saved at the last 10 epochs.

**e10s980+e3s6217+e1s6999** means epoch10step980 as the `load_chkpt` to start epoch 10\~20 training, then choose epoch3step6217 to start epoch 20\~30 training and so on.

That sounds reasonable, I will add that in the later diagram.

**Reply by lukas-blecher** [reply]
so is e3s6217 from the second run, i.e. epoch 13 effectively? 
In that case the built in validation size scheduler (`num_batches=int(args.valbatches*e/args.epochs)`) can skew the actual results.
In the early epochs the validation size is too small to come to a meaningful validation score.

https://github.com/lukas-blecher/LaTeX-OCR/blob/9f974c9c855535e1d1435236f7367cbefe6bbdda/pix2tex/train.py#L67

One option would be to add a `epoch` property in the config
```yaml
epoch: 10
```
This functionality does exist, though I didn't document or useed it aside from the option to continue training.
You'd also have to adjust the `epochs` value if it is 10.

**Reply by TITC** [reply]
noticed that but forgot, thanks for reminding me. Different num_batches make they are different validation datasets, why not set num_batches as a constant number? Is it due to early validation doesn't make much sense, so a small validation set to reduce the time spent on the validation set?

**Reply by lukas-blecher** [reply]
Yes that's it. The linear nature is probably not the best. It originally was only a quick fix.

**Reply by rainyl** [reply]
so, have you achieved higher scores?

**Reply by TITC** [reply]
@lukas-blecher Good news, Beam search achieved a higher score. 

```shell
BLEU: 0.896, ED: 7.87e-02, ACC: 0.635:  20%|█████████████████████████▎                                                                                                      | 125/633 [04:36<16:15,  1.92s/it]
```

**Reply by lukas-blecher** [reply]
Awesome!

**Reply by rainyl** [reply]
Over 0.9 now, I modified the architecture of the model with a Convnext encoder and transformer decoder 
![image](https://user-images.githubusercontent.com/30213663/171162559-90daf6c7-a183-44cc-88b3-38d04ad8da90.png)

**Reply by lukas-blecher** [reply]
Wow that's crazy. 
What kind of other changes did you make? What config did you use? 

Did you pad all images to the same size?

**Reply by TITC** [reply]
Wow!

**Reply by rainyl** [reply]
> What kind of other changes did you make? What config did you use?

TLDR: ConvNext encoder + transformer decoder

Actually, I made a new project to build and test the model, including rewrite the dataset class and dataloader using torch.data, regenerate images, then I modified the model, removed the dependency of x_transformers.

the convnext's dims are [64, 128, 256, 512], and depths are [3, 3, 9, 3], the transformer decoder's depth is 4 with dim=512, implemented by Pytorch's TransformerDecoder.

> Did you pad all images to the same size?

No, only the images in the same batch will be padding to the same size, the convnext encoder can process images with different size, but with a minimum shape of 1\*32\*32 in my configuration. 

I am still trying to improve the performance, and I will open a PR to this repo recently :)

**Reply by oakmoos** [reply]
Hello, I'd like to inquire about the last part. Could you please tell me how long you trained for? We trained for 30 epochs without loading any checkpoint files, and the accuracy remained at 0.3 without further improvement (using the default parameters from the config file). Should we adjust the training parameters or increase the number of training epochs?

**Reply by rainyl** [reply]
@oakmoos Which model did you use? My `ConvNext encoder + transformer decoder` or `vit encoder + transformer decoder` by Lukas?
If the former, maybe you could get more details at https://github.com/MODCT/Celery-LaTex-OCR
The above repo was open-sourced by us and you can get a ready-to-use app written in PySide6, totally inference locally with onnxruntime.

**Reply by oakmoos** [reply]
@rainyl  I have only trained using Lukas' model, but I'm also amazed by the results achieved with your model! I don't have much experience in training models. I used a self-generated dataset and, based on Lukas' provided weight.pth file, I was able to achieve an accuracy of over 0.85 after 10 epochs. Now, I'm exploring how much training it would take from scratch without using checkpoints to achieve a similar result. However, the results of my attempts so far haven't been too promising.I will proceed to learn more about your methodology, May I kindly ask for your guidance regarding relevant experiences?

**Comment by rainyl** [comment]
if you have enough gpu resources, you can use some hyperparameter optimization algorithm such as random search or Bayesian Model-Based optimization, there are some packages for python, such as hyperopt, more info [here](https://github.com/WillKoehrsen/hyperparameter-optimization/blob/master/Introduction%20to%20Bayesian%20Optimization%20with%20Hyperopt.ipynb)

**Reply by TITC** [reply]
Thanks for your suggestion. If you look closely, you will see that this is what I have already done in this discussion. 

> I have added 0.35 million Wikipedia data to the training set, so there are 0.5 million in total. After that, I create a [wandb sweep](https://docs.wandb.ai/guides/sweeps).

But it's kinda time consume and won't early stop. So I choose manually to shrink to a reasonable range and then use those tools.

**Reply by rainyl** [reply]
great! sorry for that, i am not familiar with wandb... I am also trying to find better hyper parameters or better model structure, but have little progress :(

**Comment by rainyl** [comment]
not much, I am trying to construct a new model, only about 0.89 bleu4 for
now, still finding better hyper parameters

Yuhang Tao ***@***.***> 于2022年5月29日周日 22:38写道：

> No, how about you?
>
> Best wishes
> Yuhang Tao
>
>
